lib/vrl/compiler/src/type_def.rs

```diff
@@ -926,6 +962,169 @@ impl TypeDef {
+                                            )]
+                                            .into_iter()
+                                        }
+                                        _ if *index < 0 => {
+                                            // If we have specified a negative index we need to merge this
+                                            // type since we aren't sure the precise index this is
+                                            // specifying since it's dependant on the array length
+                                            // at runtime.
+                                            vec![(
+                                                *idx,
+                                                kindinfo.clone().merge(
+                                                    next_segment(&mut path, kindinfo),
+                                                    false,
+                                                    true,
+                                                ),
+                                            )]
+                                            .into_iter()
+                                        }
+                                        Index::Any if !indexed => {
+                                            // If there is an Any type specified, and we know that
+                                            // there isn't a type specified for the index provided
+                                            // in the type, we need to add a specific typedef just
+                                            // for the index.
+                                            vec![
+                                                (
+                                                    Index::Index(*index as usize),
+                                                    next_segment(&mut path, kindinfo),
+                                                ),
+                                                (*idx, kindinfo.clone()),
+                                            ]
+                                            .into_iter()
+                                        }
+                                        _ => vec![(*idx, kindinfo.clone())].into_iter(),
+                                    })
+                                    .collect(),
+                            )
+                        }
+
+                        (kind, _) => kind.clone(),
+                    })
+                    .collect(),
+            ),
+        }
+    }
+
+    /// Updates the type definition at the given path with the provided typedef.
+    pub fn update_path(&self, path: &LookupBuf, type_def: &TypeDef) -> Self {
+        let segments = path.as_segments();
+        let peekable = segments.iter().peekable();
+        let kind = Self::update_segment(&self.kind, peekable, &type_def.kind);
+
+        TypeDef {
+            fallible: self.fallible,
```

Comment:
```suggestion
            fallible: type_def.fallible,
```

Reply:
I think that's questionable. The function is really only saying it will update the path, updating `fallible` is out of scope. We could rename the function `update_path_and_fallible`..

I think really, these do indicate that the parameters should be taking a `TypeKind`, even though that does mean these need exposing as `pub`.